clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java

```diff
@@ -138,103 +123,69 @@ protected StoreMetadata getStoreMetadata() {
     String resourceName = getResourceName(currentVersion);
-    long nanoTsBeforeSerialization = System.nanoTime();
-    byte[] keyBytes = keySerializer.serialize(key);
-    requestContext.requestSerializationTime = LatencyUtils.getElapsedTimeFromNSToMS(nanoTsBeforeSerialization);
-    int partitionId = metadata.getPartitionId(currentVersion, keyBytes);
+    byte[] keyBytes = requestContext.serializedKey;
+    int partitionId = requestContext.getPartitionId();
     String b64EncodedKeyBytes = EncodingUtils.base64EncodeToString(keyBytes);
 
-    requestContext.currentVersion = currentVersion;
-    requestContext.partitionId = partitionId;
-
-    String sb = URI_SEPARATOR + AbstractAvroStoreClient.TYPE_STORAGE + URI_SEPARATOR + resourceName + URI_SEPARATOR
+    uri = URI_SEPARATOR + AbstractAvroStoreClient.TYPE_STORAGE + URI_SEPARATOR + resourceName + URI_SEPARATOR
         + partitionId + URI_SEPARATOR + b64EncodedKeyBytes + AbstractAvroStoreClient.B64_FORMAT;
-    return sb;
+    requestContext.requestUri = uri;
+    return uri;
   }
 
-  private String composeRouteForBatchGetRequest(BatchGetRequestContext<K, V> requestContext) {
-    int currentVersion = getCurrentVersion();
+  private String composeURIForMultiKeyRequest(MultiKeyRequestContext<K, V> requestContext) {
+    int currentVersion = requestContext.getCurrentVersion();
     String resourceName = getResourceName(currentVersion);
 
-    requestContext.currentVersion = currentVersion;
-    StringBuilder sb = new StringBuilder();
-    sb.append(URI_SEPARATOR).append(AbstractAvroStoreClient.TYPE_STORAGE).append(URI_SEPARATOR).append(resourceName);
-    return sb.toString();
-  }
-
-  private String composeRouteForComputeRequest(ComputeRequestContext<K, V> requestContext) {
-    int currentVersion = getCurrentVersion();
-    String resourceName = getResourceName(currentVersion);
-
-    requestContext.currentVersion = currentVersion;
-    StringBuilder sb = new StringBuilder();
-    sb.append(URI_SEPARATOR).append(TYPE_COMPUTE).append(URI_SEPARATOR).append(resourceName);
-    return sb.toString();
+    RequestType requestType = requestContext.getRequestType();
+    if (requestType.equals(RequestType.MULTI_GET_STREAMING)) {
+      StringBuilder sb = new StringBuilder();
+      sb.append(URI_SEPARATOR).append(AbstractAvroStoreClient.TYPE_STORAGE).append(URI_SEPARATOR).append(resourceName);
+      return sb.toString();
+    }
+    if (requestType.equals(RequestType.COMPUTE_STREAMING)) {
+      StringBuilder sb = new StringBuilder();
+      sb.append(URI_SEPARATOR).append(TYPE_COMPUTE).append(URI_SEPARATOR).append(resourceName);
+      return sb.toString();
+    }
+    throw new VeniceClientException("Unknown request type: " + requestType);
   }
 
   private String getResourceName(int currentVersion) {
     return metadata.getStoreName() + "_v" + currentVersion;
   }
 
-  private int getCurrentVersion() {
-    int currentVersion = metadata.getCurrentStoreVersion();
```

Comment:
fetching this from metadata everytime vs making the retry context use the original context's version might cause issues during store migration where the original fails but the retry an succeed? Need to look further to see if this is an actual issue, but wanted to discuss just in case.

Reply:
`metadata` is not guaranteed to contain the most-recent info since it will be refreshed every min.
The meta point of this logic is to guarantee both the original request and the retry request will use the same store version, which is aligned with the current Router logic.

Both the original request and retry request should handle the store migration in a similar way IMO, and store migration is not supposed to delete versions.

Reply:
make sense, thanks.